Add Coretime Renewal page (#574)
polkadot-developers/polkadot-docs: develop/parachains/deployment/coretime-renewal.md

---
title: Coretime Renewal
description: TODO
---

# Coretime Renewal

## Introduction

Coretime can be purchased in bulk for a period of 28 days, providing access to Polkadot's shared security and interoperability for Polkadot Rollups (Parachains). The bulk purchase of coretime includes a rent-control mechanism that keeps future purchases within a predictable price range of the initial purchase. This allows cores to be renewed at a known price without competing against other participants in the open market.

## Bulk Sale Phases

The bulk sale process consists of three distinct phases:

- **Interlude phase** - the period between bulk sales when renewals are prioritized

> - **Leading phase** - after the interlude phase, the Coretime Chain sets a new `start_price` and initiates a Dutch auction lasting `leadin_length` blocks. During this phase, prices experience downward pressure as the system works to find the market equilibrium. The final price at the end of this phase becomes the `regular_price` that will be used in the subsequent fixed price phase

Reviewer: ```suggestion
- **Lead-in phase** - after the interlude phase, the Coretime Chain sets a new `start_price` and initiates a Dutch auction lasting `leadin_length` blocks. During this phase, prices experience downward pressure as the system works to find the market equilibrium. The final price at the end of this phase becomes the `regular_price` that will be used in the subsequent fixed price phase
```

- **Fixed price phase** - the final phase where remaining cores are sold at the `regular_price` established during the leading phase. This provides a stable and predictable pricing environment for participants who did not purchase during the price discovery period

For more comprehensive information about the coretime sales process, refer to the [Coretime Sales](https://wiki.polkadot.network/learn/learn-agile-coretime/#coretime-sales){target=\_blank} section in the Polkadot Wiki.

## Renewal Timing

> While renewals can technically be made during any phase, it is strongly recommended to complete renewals during the interlude phase. Delaying renewal introduces the risk that the core could be sold to another participant in the market, preventing successful renewal. Renewals must be initiated in advance of when you need the coretime. 

Reviewer: ```suggestion
While renewals can technically be made during any phase, it is strongly recommended to complete renewals during the interlude phase. Delaying renewal introduces the risk that the core could be sold to another participant in the market, preventing successful renewal. Renewals must be initiated well in advance to avoid the aforementioned scenario. 
```

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-1.webp)

2. Submit the `broker.renew` extrinsic
    1. Select the **broker** pallet
    2. Choose the **renew** extrinsic
    3. Fill in the **core** parameter
    4. Click the **Submit Transaction** button

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-2.webp)

For optimal results, the renewal should be performed during the interlude phase. Upon successful submission, your core will be renewed for the next coretime period, ensuring continued operation of your parachain.

## Auto Renewal

The Coretime auto-renewal feature simplifies the process of maintaining continuous coretime allocation by automatically renewing cores at the beginning of each sale period. This eliminates the need for parachains to manually renew their cores for each bulk period, reducing operational overhead and the risk of missing renewal deadlines.

When auto-renewal is enabled, the system follows this process at the start of each sale:

1. The system scans all registered auto-renewal records

> 2. For each record, it attempts to process renewal payments from the task's sovereign account

Reviewer: ```suggestion
2. For each record, it attempts to process renewal payments from the task's sovereign account (which is the sibling account on the Coretime chain derived from the Parachain ID)
```

3. Upon successful payment, the system emits a `Renewed` event and secures the core for the next period
4. If payment fails due to insufficient funds or other issues, the system emits an `AutoRenewalFailed` event

Even if an auto-renewal attempt fails, the auto-renewal setting remains active for subsequent sales. This means once you've configured auto-renewal, the setting persists across multiple periods.

> There is a limit on the total number of auto-renewals allowed, specified at the runtime level as `T::MaxAutoRenewals`.

Reviewer: ```suggestion
There is a limit on the total number of auto-renewals allowed, specified at the runtime level as `T::MaxAutoRenewals` (which is currently set to 100).
```

- Calculate it manually:

    1. Identify the appropriate prefix:

        - For sibling chains: `0x7369626c` (decodes to `b"sibl"`)
         
    2. Encode your parachain ID as a u32 [SCALE](https://docs.polkadot.com/polkadot-protocol/basics/data-encoding/#data-types){target=\_blank} value:

        - For parachain 2000, this would be `d0070000`

    3. Combine the prefix with the encoded ID to form the sovereign account address:

        - **Hex** - `0x7369626cd0070000000000000000000000000000000000000000000000000000`
        - **SS58 format** - `5Eg2fntJ27qsari4FGrGhrMqKFDRnkNSR6UshkZYBGXmSuC8`

### Auto-Renewal Configuration Extrinsics

The Coretime chain provides two primary extrinsics for managing the auto-renewal functionality:

- `enable_auto_renew(core, task, workload_end_hint)` - use this extrinsic to activate automatic renewals for a specific core. This transaction must originate from the sovereign account of the parachain task

    **Parameters:**

    - **`core`**: the core currently assigned to the task
    - **`task`**: the task for which auto-renewal is being enabled

>     - **`workload_end_hint`**: used when the core is not expiring in the current bulk period (e.g., due to a lease). This allows you to enable auto-renewal in advance, instead of needing to wait for the sale in which your core will expire

Reviewer: The recommendation is to always use this to avoid confusion. If the coretime is expiring in the current sale period, use the last timeslice number of current sale period. If it is expiring at the end of next sale period (if you purchased coretime already), then use the last timeslice number of next sale period. If there is a lease, then the timeslice when the lease ends.

Author: Suggestions applied

- **Fixed price phase** - the final phase where remaining cores are sold at the `regular_price` established during the leading phase. This provides a stable and predictable pricing environment for participants who did not purchase during the price discovery period

For more comprehensive information about the coretime sales process, refer to the [Coretime Sales](https://wiki.polkadot.network/learn/learn-agile-coretime/#coretime-sales){target=\_blank} section in the Polkadot Wiki.

## Renewal Timing

While renewals can technically be made during any phase, it is strongly recommended to complete renewals during the interlude phase. Delaying renewal introduces the risk that the core could be sold to another participant in the market, preventing successful renewal. Renewals must be initiated well in advance to avoid the aforementioned scenario. 

For example, if you purchase a core in bulk sale #1, you obtain coretime for the upcoming bulk period (during which bulk sale #2 takes place).
Your renewal must be completed during bulk sale #2, ideally during its interlude phase, to secure coretime for the subsequent period.

## Manual Renewal

> Cores can be renewed by issuing the `broker.renew(core)` extrinsic during the coretime sale period. While this process is straightforward, it requires manual action that must not be overlooked. Failure to complete this renewal step before all available cores are sold could result in your parachain being unable to secure a core for the next operational period.

Reviewer 2: ```suggestion
Cores can be renewed by issuing the [`broker.renew(core)`](https://paritytech.github.io/polkadot-sdk/master/pallet_broker/pallet/struct.Pallet.html#method.renew){target=\_blank} extrinsic during the coretime sale period. While this process is straightforward, it requires manual action that must not be overlooked. Failure to complete this renewal step before all available cores are sold could result in your parachain being unable to secure a core for the next operational period.
```

## Auto Renewal

The Coretime auto-renewal feature simplifies the process of maintaining continuous coretime allocation by automatically renewing cores at the beginning of each sale period. This eliminates the need for parachains to manually renew their cores for each bulk period, reducing operational overhead and the risk of missing renewal deadlines.

When auto-renewal is enabled, the system follows this process at the start of each sale:

1. The system scans all registered auto-renewal records
2. For each record, it attempts to process renewal payments from the task's sovereign account (which is the sibling account on the Coretime chain derived from the Parachain ID)
3. Upon successful payment, the system emits a `Renewed` event and secures the core for the next period
4. If payment fails due to insufficient funds or other issues, the system emits an `AutoRenewalFailed` event

Even if an auto-renewal attempt fails, the auto-renewal setting remains active for subsequent sales. This means once you've configured auto-renewal, the setting persists across multiple periods.

> There is a limit on the total number of auto-renewals allowed, specified at the runtime level as `T::MaxAutoRenewals` (which is currently set to 100).

Reviewer 2: ```suggestion
There is a limit on the total number of auto-renewals allowed, specified at the runtime level as [`T::MaxAutoRenewals`](https://paritytech.github.io/polkadot-sdk/master/pallet_broker/pallet/trait.Config.html#associatedtype.MaxAutoRenewals){target=\_blank} (which is currently set to 100).
```

To enable auto-renewal for your parachain, you'll need to configure several components as detailed in the following sections.

### Set Up HRMP Channel

An Horizontal Relay-routed Message Passing (HRMP) channel must be open between your parachain and the Coretime System Chain before auto-renewal can be configured. 

For instructions on establishing this connection, consult the [Opening HRMP Channels with System Parachains](/tutorials/interoperability/xcm-channels/para-to-system/){target=\_blank} guide.

### Fund Sovereign Account

The [sovereign account](https://github.com/polkadot-fellows/xcm-format/blob/10726875bd3016c5e528c85ed6e82415e4b847d7/README.md?plain=1#L50){target=\_blank} of your parachain on the Coretime chain needs adequate funding to cover both XCM transaction fees and the recurring coretime renewal payments.

To determine your parachain's sovereign account address, you can:

- Use the "Para ID" to Address section in [Substrate Utilities](https://www.shawntabrizi.com/substrate-js-utilities/){target=\_blank} with the **Sibling** option selected

- Calculate it manually:

    1. Identify the appropriate prefix:

>         - For sibling chains: `0x7369626c` (decodes to `b"sibl"`)

Reviewer 2: ```suggestion
        - For sibling chains - `0x7369626c` (decodes to `b"sibl"`)
```

         
    2. Encode your parachain ID as a u32 [SCALE](https://docs.polkadot.com/polkadot-protocol/basics/data-encoding/#data-types){target=\_blank} value:

        - For parachain 2000, this would be `d0070000`

    3. Combine the prefix with the encoded ID to form the sovereign account address:

        - **Hex** - `0x7369626cd0070000000000000000000000000000000000000000000000000000`
        - **SS58 format** - `5Eg2fntJ27qsari4FGrGhrMqKFDRnkNSR6UshkZYBGXmSuC8`

### Auto-Renewal Configuration Extrinsics

The Coretime chain provides two primary extrinsics for managing the auto-renewal functionality:

> - `enable_auto_renew(core, task, workload_end_hint)` - use this extrinsic to activate automatic renewals for a specific core. This transaction must originate from the sovereign account of the parachain task

Reviewer 2: ```suggestion
- [`enable_auto_renew(core, task, workload_end_hint)`](https://paritytech.github.io/polkadot-sdk/master/pallet_broker/pallet/struct.Pallet.html#method.enable_auto_renew){target=\_blank} - use this extrinsic to activate automatic renewals for a specific core. This transaction must originate from the sovereign account of the parachain task
```

## Manual Renewal

Cores can be renewed by issuing the [`broker.renew(core)`](https://paritytech.github.io/polkadot-sdk/master/pallet_broker/pallet/struct.Pallet.html#method.renew){target=\_blank} extrinsic during the coretime sale period. While this process is straightforward, it requires manual action that must not be overlooked. Failure to complete this renewal step before all available cores are sold could result in your parachain being unable to secure a core for the next operational period.

To manually renew a core:

1. In [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=\_blank}, connect to the Coretime chain, navigate to the **Developer** dropdown and select the **Extrinsics** option

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-1.webp)

2. Submit the `broker.renew` extrinsic
    1. Select the **broker** pallet
    2. Choose the **renew** extrinsic
    3. Fill in the **core** parameter
    4. Click the **Submit Transaction** button

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-2.webp)

> For optimal results, the renewal should be performed during the interlude phase. Upon successful submission, your core will be renewed for the next coretime period, ensuring continued operation of your parachain.

Reviewer 3: Using parachain here instead of rollup, not sure if that was intended so just wanted to flag it

- **`core`** - identify which core your parachain is assigned to when the it expires. This requires checking both current assignments and planned future assignments:
    - For current period - query `broker.workload()`
    - For next period - query `broker.workplan()`

    **Example for parachain `2000`:**
    
    Current assignment (workload)
    ```
    [
      [50]
      [{
        mask: 0xffffffffffffffffffff
        assignment: {Task: 2,000}
      }]
    ]
    ```

    Future assignment (workplan)
    ```
    [
      [[322,845, 48]]
      [{
        mask: 0xffffffffffffffffffff
        assignment: {Task: 2,000}
      }]
    ]
    ```

    **Note:** use the core from workplan (`48` in this example) if your task appears there. Only use the core from workload if it's not listed in workplan.

- **`task`** - use your parachain ID, which can be verified by connecting to your parachain and querying `parachainInfo.parachainId()`

- **`workload_end_hint`** - you should always set it explicitly to avoid misbehavior. This value indicates when your assigned core will expire. Here's how to calculate the correct value based on how your core is assigned:

>     - If the parachain uses bulk coretime:
> 
>         Query `broker.saleinfo`. You’ll get a result like:

Reviewer 3: ```suggestion
    - If the parachain uses bulk coretime, query `broker.saleinfo`. You’ll get a result like:
```

        - If the core has already been renewed and will expire in the next sale, use the `regionEnd` value. In this example, that would be `327885`

    - If the parachain has a lease:

        Query `broker.leases`, which returns entries like:

        ```json
        [
          {
            "until": 359280,
            "task": 2035
          },
          ...
        ]
        ```

        - Use the `until` value of the lease corresponding to your task. For example, `359280` would be the value for `workload_end_hint` in the case of task `2035`

Once you have these values, construct the extrinsic:

1. In [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=\_blank}, connect to the Coretime chain, navigate to the **Developer** dropdown and select the **Extrinsics** option

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-1.webp)

2. Create the `broker.enable_auto_renew` extrinsic
    1. Select the **broker** pallet
    2. Choose the **enableAutoRenew** extrinsic
    3. Fill in the parameters
    4. Copy the encoded call data

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-3.webp)

>     For parachain `2000` on core `48` with `workload_end_hint` `327885`, the encoded call data would be:`0x32153000d007000001cd000500`

Reviewer 3: ```suggestion
    For parachain `2000` on core `48` with `workload_end_hint` `327885`, the **encoded call data** is:`0x32153000d007000001cd000500`
```

2. Create a `sudo.sudo` extrinsic that executes `polkadotXcm.send`:
    1. Use the `sudo.sudo` extrinsic to execute the following call as Root
    2. Select the **polkadotXcm** pallet
    3. Choose the **send** extrinsic
    4. Set the **dest** parameter as the Coretime chain (Parachain 1005)

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-5.webp)

3. Construct the XCM and submit it:

    1. Add a **WithdrawAsset** instruction
    2. Add a **BuyExecution** instruction
    3. Add a **Transact** instruction with the following parameters:
        - **originKind** - use `SovereignAccount`
        - **requireWeightAtMost** - use the weight calculated previously
        - **call** - use the encoded call data generated before
    4. Add a **RefundSurplus** instruction
    5. Add a **DepositAsset** instruction to send remaining funds to the parachain sovereign account
    6. Click the **Submit Transaction** button

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-6.webp)

> After successful execution, your parachain should have auto-renewal enabled. To verify this, check the events emitted in the Coretime chain. You should see confirmation events similar to:

Reviewer 3: Are you specifically talking about the `broker.AutoRenewalEnabled`? If so, you should put it in text instead of solely relying on the image (for accessibility purposes)

## Renewal Timing

While renewals can technically be made during any phase, it is strongly recommended that they be completed during the interlude phase. Delaying renewal introduces the risk that the core could be sold to another market participant, preventing successful renewal. Renewals must be initiated well in advance to avoid the scenario above. 

For example, if you purchase a core in bulk sale #1, you obtain coretime for the upcoming bulk period (during which bulk sale #2 takes place).
Your renewal must be completed during bulk sale #2, ideally during its interlude phase, to secure coretime for the subsequent period.

## Manual Renewal

Cores can be renewed by issuing the [`broker.renew(core)`](https://paritytech.github.io/polkadot-sdk/master/pallet_broker/pallet/struct.Pallet.html#method.renew){target=\_blank} extrinsic during the coretime sale period. While this process is straightforward, it requires manual action that must not be overlooked. Failure to complete this renewal step before all available cores are sold could result in your parachain being unable to secure a core for the next operational period.

To manually renew a core:

> 1. In [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=\_blank}, connect to the Coretime chain, navigate to the **Developer** dropdown and select the **Extrinsics** option

Reviewer 4: ```suggestion
1. In [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=\_blank}, connect to the Coretime chain, navigate to the **Developer** dropdown, and select the **Extrinsics** option
```

When auto-renewal is enabled, the system follows this process at the start of each sale:

1. The system scans all registered auto-renewal records
2. For each record, it attempts to process renewal payments from the task's [sovereign account](/polkadot-protocol/glossary/#sovereign-account){target=\_blank} (which is the sibling account on the Coretime chain derived from the parachain ID)
3. Upon successful payment, the system emits a `Renewed` event and secures the core for the next period
4. If payment fails due to insufficient funds or other issues, the system emits an `AutoRenewalFailed` event

Even if an auto-renewal attempt fails, the auto-renewal setting remains active for subsequent sales. This means the setting persists across multiple periods once you've configured auto-renewal.

To enable auto-renewal for your parachain, you must configure several components, as detailed in the following sections.

### Set Up an HRMP Channel

> A Horizontal Relay-routed Message Passing (HRMP) channel must be opened between your parachain and the Coretime system chain before auto-renewal can be configured. 

Reviewer 4: ```suggestion
A Horizontal Relay-routed Message Passing (HRMP) channel must be opened between your parachain and the Coretime system chain before configuring auto-renewal. 
```

        - If the coretime expires in the current sale period, use the last timeslice of the current sale period

        - If the coretime expires at the end of the next sale period (e.g., because you've already renewed), use the last timeslice of the next sale period

        - If a lease is active, use the timeslice when the lease ends

- [**`disable_auto_renew(core, task)`**](https://paritytech.github.io/polkadot-sdk/master/pallet_broker/pallet/struct.Pallet.html#method.disable_auto_renew){target=\_blank} - use this extrinsic to stop automatic renewals. This extrinsic also requires that the origin is the sovereign account of the parachain task

     **Parameters:**

    - **`core`** - the core currently assigned to the task
    - **`task`** - the task for which auto-renewal is enabled

### Construct the Enable Auto-Renewal Extrinsic

To configure auto-renewal, you'll need to gather specific information for the `enable_auto_renew` extrinsic parameters:

> - **`core`** - identify which core your parachain is assigned to when the it expires. This requires checking both current assignments and planned future assignments:

Reviewer 4: ```suggestion
- **`core`** - identify which core your parachain is assigned to when it expires. This requires checking both current assignments and planned future assignments:
```

- **`workload_end_hint`** - you should always set it explicitly to avoid misbehavior. This value indicates when your assigned core will expire. Here's how to calculate the correct value based on how your core is assigned:
    - If the parachain uses bulk coretime, query `broker.saleinfo`. You’ll get a result like:

        ```json
        {
        "saleStart": 1544949,
        "leadinLength": 100800,
        "endPrice": 922760076,
        "regionBegin": 322845,
        "regionEnd": 327885,
        "idealCoresSold": 18,
        "coresOffered": 18,
        "firstCore": 44,
        "selloutPrice": 92272712073,
        "coresSold": 18
        }
        ```

        - If the core expires in the current sale, use the `regionBegin` value, which in this case is  `322845`

        - If the core has already been renewed and will expire in the next sale, use the `regionEnd` value. In this example, that would be `327885`

    - If the parachain has a lease, query `broker.leases`, which returns entries like:

        ```json
        [
          {
            "until": 359280,
            "task": 2035
          },
          ...
        ]
        ```

>         - Use the `until` value of the lease corresponding to your task. For example, `359280` would be the value for `workload_end_hint` in the case of task `2035`

Reviewer 4: ```suggestion
        - Use the `until` value of the lease that corresponds to your task. For example, `359280` would be the value for `workload_end_hint` in the case of task `2035`
```

Once you have these values, construct the extrinsic:

> 1. In [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=\_blank}, connect to the Coretime chain, navigate to the **Developer** dropdown and select the **Extrinsics** option

Reviewer 4: ```suggestion
1. In [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=\_blank}, connect to the Coretime chain, navigate to the **Developer** dropdown, and select the **Extrinsics** option
```

3. Execute the auto-renewal extrinsic
4. Refund surplus DOT back to the sovereign account

Here's how to submit this XCM using Acala (Parachain 2000) as an example:

1. In [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=\_blank}, connect to your parachain, navigate to the **Developer** dropdown and select the **Extrinsics** option

2. Create a `sudo.sudo` extrinsic that executes `polkadotXcm.send`:
    1. Use the `sudo.sudo` extrinsic to execute the following call as Root
    2. Select the **polkadotXcm** pallet
    3. Choose the **send** extrinsic
    4. Set the **dest** parameter as the Coretime chain (Parachain 1005)

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-5.webp)

3. Construct the XCM and submit it:

    1. Add a **WithdrawAsset** instruction
    2. Add a **BuyExecution** instruction
    3. Add a **Transact** instruction with the following parameters:
        - **originKind** - use `SovereignAccount`
        - **requireWeightAtMost** - use the weight calculated previously
        - **call** - use the encoded call data generated before
    4. Add a **RefundSurplus** instruction

>     5. Add a **DepositAsset** instruction to send remaining funds to the parachain sovereign account

Reviewer 4: ```suggestion
    5. Add a **DepositAsset** instruction to send the remaining funds to the parachain sovereign account
```

---
title: Coretime Renewal
description: Learn how to renew coretime manually or automatically to ensure uninterrupted parachain operation with predictable pricing and minimal risk.
---

# Coretime Renewal

## Introduction

Coretime can be purchased in bulk for a period of 28 days, providing access to Polkadot's shared security and interoperability for Polkadot Rollups (Parachains). The bulk purchase of coretime includes a rent-control mechanism that keeps future purchases within a predictable price range of the initial purchase. This allows cores to be renewed at a known price without competing against other participants in the open market.

## Bulk Sale Phases

The bulk sale process consists of three distinct phases:

- **Interlude phase** - the period between bulk sales when renewals are prioritized

> - **Lead-in phase** - after the interlude phase, the Coretime Chain sets a new `start_price` and initiates a Dutch auction lasting `leadin_length` blocks. During this phase, prices experience downward pressure as the system works to find the market equilibrium. The final price at the end of this phase becomes the `regular_price` that will be used in the subsequent fixed price phase

Reviewer 5: Why is the Coretime Chain relevant here?

- **Fixed price phase** - the final phase where remaining cores are sold at the `regular_price` established during the leading phase. This provides a stable and predictable pricing environment for participants who did not purchase during the price discovery period

> For more comprehensive information about the coretime sales process, refer to the [Coretime Sales](https://wiki.polkadot.network/learn/learn-agile-coretime/#coretime-sales){target=\_blank} section in the Polkadot Wiki.

Reviewer 5: Use a period at the end of a sentence...

## Renewal Timing

While renewals can technically be made during any phase, it is strongly recommended to complete renewals during the interlude phase. Delaying renewal introduces the risk that the core could be sold to another participant in the market, preventing successful renewal. Renewals must be initiated well in advance to avoid the aforementioned scenario. 

For example, if you purchase a core in bulk sale #1, you obtain coretime for the upcoming bulk period (during which bulk sale #2 takes place).
Your renewal must be completed during bulk sale #2, ideally during its interlude phase, to secure coretime for the subsequent period.

## Manual Renewal

Cores can be renewed by issuing the [`broker.renew(core)`](https://paritytech.github.io/polkadot-sdk/master/pallet_broker/pallet/struct.Pallet.html#method.renew){target=\_blank} extrinsic during the coretime sale period. While this process is straightforward, it requires manual action that must not be overlooked. Failure to complete this renewal step before all available cores are sold could result in your parachain being unable to secure a core for the next operational period.

To manually renew a core:

> 1. In [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=\_blank}, connect to the Coretime chain, navigate to the **Developer** dropdown and select the **Extrinsics** option

Reviewer 5: Why did you capitalize "chain" earlier (L17) but not here? Please be consistent in style.

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-1.webp)

2. Submit the `broker.renew` extrinsic
    1. Select the **broker** pallet
    2. Choose the **renew** extrinsic
    3. Fill in the **core** parameter
    4. Click the **Submit Transaction** button

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-2.webp)

For optimal results, the renewal should be performed during the interlude phase. Upon successful submission, your core will be renewed for the next coretime period, ensuring continued operation of your parachain.

## Auto Renewal

> The Coretime auto-renewal feature simplifies the process of maintaining continuous coretime allocation by automatically renewing cores at the beginning of each sale period. This eliminates the need for parachains to manually renew their cores for each bulk period, reducing operational overhead and the risk of missing renewal deadlines.

Reviewer 5: Again, inconsistent capitalization on "Coretime".

When auto-renewal is enabled, the system follows this process at the start of each sale:

1. The system scans all registered auto-renewal records

> 2. For each record, it attempts to process renewal payments from the task's sovereign account (which is the sibling account on the Coretime chain derived from the Parachain ID)

Reviewer 5: Should link to something about sibling account derivation, otherwise it's confusing.

Author: Do you have a resource in mind that you'd suggest linking to?

3. Upon successful payment, the system emits a `Renewed` event and secures the core for the next period
4. If payment fails due to insufficient funds or other issues, the system emits an `AutoRenewalFailed` event

Even if an auto-renewal attempt fails, the auto-renewal setting remains active for subsequent sales. This means once you've configured auto-renewal, the setting persists across multiple periods.

There is a limit on the total number of auto-renewals allowed, specified at the runtime level as [`T::MaxAutoRenewals`](https://paritytech.github.io/polkadot-sdk/master/pallet_broker/pallet/trait.Config.html#associatedtype.MaxAutoRenewals){target=\_blank} (which is currently set to 100).

To enable auto-renewal for your parachain, you'll need to configure several components as detailed in the following sections.

### Set Up HRMP Channel

> An Horizontal Relay-routed Message Passing (HRMP) channel must be open between your parachain and the Coretime System Chain before auto-renewal can be configured. 

Reviewer 5: ```suggestion
A Horizontal Relay-routed Message Passing (HRMP) channel must be open between your parachain and the Coretime System Chain before auto-renewal can be configured. 
```

For instructions on establishing this connection, consult the [Opening HRMP Channels with System Parachains](/tutorials/interoperability/xcm-channels/para-to-system/){target=\_blank} guide.

### Fund Sovereign Account

The [sovereign account](https://github.com/polkadot-fellows/xcm-format/blob/10726875bd3016c5e528c85ed6e82415e4b847d7/README.md?plain=1#L50){target=\_blank} of your parachain on the Coretime chain needs adequate funding to cover both XCM transaction fees and the recurring coretime renewal payments.

To determine your parachain's sovereign account address, you can:

- Use the "Para ID" to Address section in [Substrate Utilities](https://www.shawntabrizi.com/substrate-js-utilities/){target=\_blank} with the **Sibling** option selected

> - Calculate it manually:

Reviewer 5: Ugh, we need proper tools for this (cc @acatangiu)

        - For sibling chains - `0x7369626c` (decodes to `b"sibl"`)
         
    2. Encode your parachain ID as a u32 [SCALE](https://docs.polkadot.com/polkadot-protocol/basics/data-encoding/#data-types){target=\_blank} value:

        - For parachain 2000, this would be `d0070000`

    3. Combine the prefix with the encoded ID to form the sovereign account address:

        - **Hex** - `0x7369626cd0070000000000000000000000000000000000000000000000000000`
        - **SS58 format** - `5Eg2fntJ27qsari4FGrGhrMqKFDRnkNSR6UshkZYBGXmSuC8`

### Auto-Renewal Configuration Extrinsics

The Coretime chain provides two primary extrinsics for managing the auto-renewal functionality:

- [`enable_auto_renew(core, task, workload_end_hint)`](https://paritytech.github.io/polkadot-sdk/master/pallet_broker/pallet/struct.Pallet.html#method.enable_auto_renew){target=\_blank} - use this extrinsic to activate automatic renewals for a specific core. This transaction must originate from the sovereign account of the parachain task

    **Parameters:**

    - **`core`** - the core currently assigned to the task
    - **`task`** - the task for which auto-renewal is being enabled

>     - **`workload_end_hint`** - the timeslice at which the currently assigned core will stop being used. This value helps the system determine when auto-renewal should begin. It is recommended to always provide this value to avoid ambiguity

Reviewer 5: If you use complete sentences then you need to be consistent across bullet points. As in, the last sentence is not exempt from a period at the end.

    - **`task`**: the task for which auto-renewal is enabled

### Construct the Enable Auto Renewal Extrinsic

To configure auto-renewal, you'll need to gather specific information for the `enable_auto_renew` extrinsic parameters:

- **`core`** - identify which core your parachain is assigned to when the it expires. This requires checking both current assignments and planned future assignments:
    - For current period - query `broker.workload()`
    - For next period - query `broker.workplan()`

    **Example for parachain `2000`:**
    
    Current assignment (workload)
    ```
    [
      [50]
      [{
        mask: 0xffffffffffffffffffff
        assignment: {Task: 2,000}
      }]
    ]
    ```

    Future assignment (workplan)
    ```
    [
      [[322,845, 48]]
      [{
        mask: 0xffffffffffffffffffff
        assignment: {Task: 2,000}
      }]
    ]
    ```

    **Note:** use the core from workplan (`48` in this example) if your task appears there. Only use the core from workload if it's not listed in workplan.

- **`task`** - use your parachain ID, which can be verified by connecting to your parachain and querying `parachainInfo.parachainId()`

> - **`workload_end_hint`** - you should always set it explicitly to avoid misbehavior. This value indicate when your assigned core will expire. Here's how to calculate the correct value based on how your core is assigned:

Reviewer 5: ```suggestion
- **`workload_end_hint`** - you should always set it explicitly to avoid misbehavior. This value indicates when your assigned core will expire. Here's how to calculate the correct value based on how your core is assigned:
```

    - If the parachain uses bulk coretime:

        Query `broker.saleinfo`. You’ll get a result like:

        ```json
        {
        "saleStart": 1544949,
        "leadinLength": 100800,
        "endPrice": 922760076,
        "regionBegin": 322845,
        "regionEnd": 327885,
        "idealCoresSold": 18,
        "coresOffered": 18,
        "firstCore": 44,
        "selloutPrice": 92272712073,
        "coresSold": 18
        }
        ```

        - If the core expires in the current sale, use the `regionBegin` value, which in this case is  `322845`

        - If the core has already been renewed and will expire in the next sale, use the `regionEnd` value. In this example, that would be `327885`

    - If the parachain has a lease:

        Query `broker.leases`, which returns entries like:

        ```json
        [
          {
            "until": 359280,
            "task": 2035
          },
          ...
        ]
        ```

>         - Use the `until` value of the lease corresponding to your task. For example, `359280` would be the value for `workload_end_hint` in the case of task `2035`

Reviewer 5: ```suggestion
        - Use the `until` value of the lease corresponding to your task. For example, `359280` would be the value for `workload_end_hint` in the case of task `2035`.
```

Once you have these values, construct the extrinsic:

> 1. In [Polkadot.js Apps](https://polkadot.js.org/apps/#/explorer){target=\_blank}, connect to the Coretime chain, navigate to the **Developer** dropdown and select the **Extrinsics** option

Reviewer 5: Why don't you just link to the Coretime Chain / Extrinsics tab here?

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-1.webp)

2. Create the `broker.enable_auto_renew` extrinsic
    1. Select the **broker** pallet
    2. Choose the **enableAutoRenew** extrinsic
    3. Fill in the parameters
    4. Copy the encoded call data

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-3.webp)

>     For parachain `2000` on core `48` with workload_end_hint `327885`, the encoded call data would be:`0x32153000d007000001cd000500`

Reviewer 5: ```suggestion
    For parachain `2000` on core `48` with `workload_end_hint` `327885`, the encoded call data would be:`0x32153000d007000001cd000500`
```

3. Check the transaction weight for executing the call. You can estimate this by executing the `transactionPaymentCallApi.queryCallInfo` runtime call with the encoded call data previously obtained:

    ![](/images/develop/parachains/deployment/coretime-renewal/coretime-renewal-4.webp)

### Submit the XCM from your Parachain

To activate auto-renewal, you must submit an XCM from your parachain to the Coretime chain using Root origin. This can be done either through the sudo pallet (if available) or through your parachain's governance system.

> The XCM message needs to execute these operations:

Reviewer 5: Cross-Consensus Message Message?
